File: tasty4scalac/integration/test/tasty4scalac/CompilerTest.java
package tasty4scalac;

import org.junit.Assert;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.junit.runners.Parameterized;
import scala.collection.immutable.Set;

import java.util.Arrays;
import java.util.Collection;

@RunWith(Parameterized.class)
public final class CompilerTest {
    private final Compiler compiler;

    public CompilerTest(Compiler factory) {
        this.compiler = factory;
    }

    @Parameterized.Parameters(name = "{index}: {0}")
    public static Collection<Compiler[]> parameters() {
        return Arrays.asList(new Compiler[][]{
                {Scalac$.MODULE$.apply()},
                {Dotty$.MODULE$.apply()}
        });
    }

    @Test
    public void generatesTasty() {
        String code = "class A";
        Set<Tasty> tastySet = compiler.compile(code);
        Assert.assertFalse(tastySet.isEmpty());
    }
}

File: tasty4scalac/integration/test/tasty4scalac/CompilerTest.scala
package tasty4scalac

import java.nio.file.{Files, Path}

import org.junit.Assert._
import org.junit.Test
import org.junit.runner.RunWith
import org.junit.runners.Parameterized
import org.junit.runners.Parameterized.Parameters
import tasty.binary.BinaryInput
import tasty.{TastyADT, TastyADTUnpickler}

@RunWith(classOf[Parameterized])
class CompilerTest(fileName: String, scalacOutput: Map[String, TastyADT], dottyOutput: Map[String, TastyADT]) {

  @Test
  def scalacEmitsSameAmountOfTastyFiles(): Unit =
    assertEquals(dottyOutput.keySet, scalacOutput.keySet)

  @Test
  def scalacEmitsSubsetOfDottyEmittedNames(): Unit = {
    for {
      (name, scalacTasty) <- scalacOutput
      dottyTasty = dottyOutput(name)
      scalacEmittedNames = scalacTasty.names.toSet
      dottyEmittedNames = dottyTasty.names.toSet
      missingNames = scalacEmittedNames -- dottyEmittedNames
    } assertTrue(s"scalac emitted non-matching names for: $name. Those are: [$missingNames]", missingNames.isEmpty)
  }
}

object CompilerTest {

  import collection.JavaConverters._

  @Parameters(name = "{0}")
  def data(): java.util.Collection[Array[Any]] = {
    val testCases = for {
      testCase <- ScalaTestCaseSource.testCases()
      name = testCase.getFileName.toString
      code = readContent(testCase)
      scalacOutput = Scalac().compile(code).mapValues(convert)
      dottyOutput = Dotty().compile(code).mapValues(convert)
    } yield Array[Any](name, scalacOutput, dottyOutput)

    testCases.asJavaCollection
  }

  private def readContent(path: Path): String = new String(Files.readAllBytes(path))

  private def convert(binary: BinaryTasty): TastyADT = {
    val input = new BinaryInput(binary.bytes)
    TastyADTUnpickler.unpickle(input)
  }
}
